pkg/core/native/native_test/cryptolib_verification_test.go

```diff
@@ -0,0 +1,354 @@
+		// takes as an argument unhashed message and then takes sha256 hash of this message.
+		signature := pk.SignHash(hash.Sha256(msg))
+
+		// Ensure that signature verification passes. This line here is just for testing purposes,
+		// it won't be present in the real code.
+		require.True(t, pk.PublicKey().Verify(signature, hash.Sha256(msg).BytesBE()))
+
+		// Build invocation witness script for the user's account.
+		invBytes := buildKoblitzInvocationScript(t, signature)
+
+		// Construct witness for signer #0 (the user itself).
+		tx.Scripts = []transaction.Witness{
+			{
+				InvocationScript:   invBytes,
+				VerificationScript: vrfBytes,
+			},
+		}
+
+		// Add transaction to the chain. No error is expected on new block addition. Note, that this line performs
+		// all those checks that are executed during transaction acceptance in the real network.
+		e.AddNewBlock(t, tx)
+
+		// Double-check: ensure funds have been transferred.
+		e.CheckGASBalance(t, to, big.NewInt(int64(amount)))
+	}
+
+	// N3 compatible witness verification script has larger constant length and higher execution cost
+	// (186 bytes, 5116020 GAS including Invocation script execution).
+	check(t, buildKoblitzVerificationScriptCompat, constructMessageCompat)
+	// N3 incompatible witness verification script has lower variable length and lower execution cost
+	// (136 bytes, 4120620 GAS including Invocation script execution).
+	check(t, buildKoblitzVerificationScriptSimple, constructMessageSimple)
+}
+
+// buildKoblitzVerificationScript builds custom verification script for the provided Koblitz public key.
+// It checks that the following message is signed by the provided public key:
+//
+//	sha256([4-bytes-network-magic-LE, txHash-bytes-BE])
+//
+// It produces constant-length verification script (186 bytes) independently of the network parameters.
+func buildKoblitzVerificationScriptCompat(t *testing.T, pub *keys.PublicKey) []byte {
+	criptoLibH := state.CreateNativeContractHash(nativenames.CryptoLib)
+
+	// vrf is witness verification script corresponding to the pub.
+	vrf := io.NewBufBinWriter()
+	emit.Int(vrf.BinWriter, int64(native.Secp256k1)) // push Koblitz curve identifier.
+	emit.Opcodes(vrf.BinWriter, opcode.SWAP)         // swap curve identifier with the signature.
+	emit.Bytes(vrf.BinWriter, pub.Bytes())           // emit the caller's public key.
+	// Construct and push the signed message. The signed message is effectively the network-dependent transaction hash,
+	// i.e. msg = Sha256([4-bytes-network-magic-LE, tx.Hash()])
+	// Firstly, convert network magic (uint32) to LE buffer.
+	emit.Syscall(vrf.BinWriter, interopnames.SystemRuntimeGetNetwork) // push network magic.
+	// First byte: n & 0xFF
+	emit.Opcodes(vrf.BinWriter, opcode.DUP)
+	emit.Int(vrf.BinWriter, 0xFF) // TODO: this can be optimize in order not to allocate 0xFF every time, but need to compare execution price.
```

Comment:
SUBSTR/LEFT can also be handy. But at the same time any of them likely will require conditionals/loops while your code doesn't.

Reply:
Maybe something like

GetNetwork  // int on the stack, can have 0-5 bytes length serialized
PUSHINT64 0x100000000
ADD // Some new number that is 5 bytes at least when serialized, but first 4 bytes are intact
PUSH4
LEFT // Now we cut the first 4 bytes out of a number that is at least 5 bytes long, it will succeed and it doesn't need cycles